blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,958 @@
+	// Main Object:
+	// 		kvstore: persistent storage, used to initialize index cache at startup
+	// 		cache: in-memory index for clean data, used to query index data
+	//      dirty: the cache to update during event processing, will be merged to clean cache after all events are processed. If errors occur during event processing, dirty cache will be discarded.
+	liquidStakingIndexer struct {
+		kvstore       db.KVStore                      // persistent storage
+		cache         *liquidStakingCacheThreadSafety // in-memory index for clean data
+		blockInterval time.Duration
+	}
+)
+
+var (
+	_liquidStakingInterface           abi.ABI
+	_liquidStakingHeightKey           = []byte("lsHeight")
+	_liquidStakingTotalBucketCountKey = []byte("lsTotalBucketCount")
+
+	errBucketTypeNotExist = errors.New("bucket type does not exist")
+
+	// ErrBucketInfoNotExist is the error when bucket does not exist
+	ErrBucketInfoNotExist = errors.New("bucket info does not exist")
+)
+
+func init() {
+	var err error
+	_liquidStakingInterface, err = abi.JSON(strings.NewReader(LiquidStakingContractABI))
+	if err != nil {
+		panic(err)
+	}
+}
+
+// NewLiquidStakingIndexer creates a new liquid staking indexer
+func NewLiquidStakingIndexer(kvStore db.KVStore, blockInterval time.Duration) LiquidStakingIndexer {
+	return &liquidStakingIndexer{
+		blockInterval: blockInterval,
+		kvstore:       kvStore,
+		cache:         newLiquidStakingCache(),
+	}
+}
+
+// Start starts the indexer
+func (s *liquidStakingIndexer) Start(ctx context.Context) error {
+	if err := s.kvstore.Start(ctx); err != nil {
+		return err
+	}
+	return s.loadCache()
+}
+
+// Stop stops the indexer
+func (s *liquidStakingIndexer) Stop(ctx context.Context) error {
+	if err := s.kvstore.Stop(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// PutBlock puts a block into indexer
+func (s *liquidStakingIndexer) PutBlock(ctx context.Context, blk *block.Block) error {
+	// new dirty cache for this block
+	// it's not necessary to use thread safe cache here, because only one thread will call this function
+	// and no update to cache will happen before dirty merge to clean
+	dirty := newLiquidStakingDirty(s.cache.cache)
+	dirty.putHeight(blk.Height())
+
+	// make action map
+	actionMap := make(map[hash.Hash256]*action.SealedEnvelope)
+	for i := range blk.Actions {
+		h, err := blk.Actions[i].Hash()
+		if err != nil {
+			return err
+		}
+		actionMap[h] = &blk.Actions[i]
+	}
+
+	// handle events of block
+	for _, receipt := range blk.Receipts {
+		if receipt.Status != uint64(iotextypes.ReceiptStatus_Success) {
+			continue
+		}
+		act, ok := actionMap[receipt.ActionHash]
+		if !ok {
+			return errors.Errorf("action %x not found", receipt.ActionHash)
+		}
+		for _, log := range receipt.Logs() {
+			if log.Address != LiquidStakingContractAddress {
+				continue
+			}
+			if err := s.handleEvent(ctx, dirty, blk, act, log); err != nil {
```

Comment:
if we move this to the line above `s.commit(dirty)`, is `dirty` still necessary?